src/WorkItemMigrator/JiraExport/JiraProvider.cs

```diff
@@ -405,25 +395,40 @@ public string GetUserEmail(string usernameOrAccountId)
         public string GetCustomId(string propertyName)
         {
             var customId = string.Empty;
-            var response = (JArray)Jira.RestClient.ExecuteRequestAsync(Method.GET, $"{JiraApiV2}/field").Result;
-            foreach (var item in response)
+            if (JiraNameFieldCache == null)
             {
-                var nameField = (JValue)item.SelectToken("name");
-                if (nameField.Value.ToString().ToLower() == propertyName.ToLower())
-                {
-                    var idField = (JValue)item.SelectToken("id");
-                    customId = idField.Value.ToString();
-                }
+                var response = (JArray)_jiraServiceWrapper.RestClient.ExecuteRequestAsync(Method.GET, $"{JiraApiV2}/field").Result;
+                JiraNameFieldCache = response
+                    .Where(x => x.Value<string>("name") != null && x.Value<string>("id") != null)
+                    .Select(x => new { name = x.Value<string>("name").ToLower(), id = x.Value<string>("id").ToLower() })
+                    .ToLookup(l => l.name, l => l.id);
+
+                JiraKeyFieldCache = response
+                    .Where(x => x.Value<string>("key") != null && x.Value<string>("id") != null)
+                    .Select(x => new { key = x.Value<string>("key").ToLower(), id = x.Value<string>("id").ToLower() })
+                    .ToLookup(l => l.key, l => l.id);
+            }
+
+            customId = GetItemFromFieldCache(propertyName, JiraNameFieldCache);
+
+            if (string.IsNullOrEmpty(customId))
+            {
+                customId = GetItemFromFieldCache(propertyName, JiraKeyFieldCache);
+            }
+
+            return customId;
+        }
 
-                if (string.IsNullOrEmpty(customId))
+        private string GetItemFromFieldCache(string propertyName, ILookup<string,string> cache)
+        {
+            string customId = null;
+            var query = cache.Where(x => x.Key.Equals(propertyName.ToLower())).FirstOrDefault();
+            if (query != null)
+            {
+                customId = query.Count() > 0 ? query.First() : null;
```

Comment:
How do you know if the first one is correct if you get multiple hits in the grouping ?

Reply:
This was the same behavior as we had before, in the case where we received multiple fields with the same key. We added a warning statement in can we have multiple cache hits

Reply:
@madkoo , @moerketh , what is the status here? :)

Reply:
Ok, let it be then.